src/main/java/io/github/syst3ms/skriptparser/effects/EffWait.java

```diff
@@ -44,15 +45,14 @@ protected void execute(TriggerContext ctx) {
     }
 
     @Override
-    public Statement walk(TriggerContext ctx) {
-        Duration dur = duration.getSingle(ctx);
-        if (dur == null)
+    public Optional<? extends Statement> walk(TriggerContext ctx) {
+        Optional<? extends Duration> dur = duration.getSingle(ctx);
+        if (dur.isEmpty())
             return getNext();
-        if (getNext() == null)
-            return null;
-
-        ThreadUtils.runAfter(() -> Statement.runAll(getNext(), ctx), dur);
-        return null;
+        if (getNext().isEmpty())
+            return Optional.empty();
+        ThreadUtils.runAfter(() -> Statement.runAll(getNext().get(), ctx), dur.get());
+        return Optional.empty();
```

Comment:
I love this overall PR: it adds consistency and clarity about the former frustrating `getSingle(TriggerContext)` method and others. But for me, this is a bit 'over the top'.
I think your main goal was to change as many Nullable methods with Optionals, and that's great, but I think the change inside of the actual Statement class (and its subclasses) is not necessary? I get what you want to achieve with this, but this adds a lot of boilerplate code (which was formerly present as well under the form of nullity checks).

It's a 'double-edged' sword for me. It adds so much utility, but if you're going to implement it everywhere, I think it's a bit too much. That said, having both Nullable values and Optionals may be confusing as well. I don't know what I have to think about this honestly. It will probably become a habit once it's implemented?

Reply:
Having both nullable values and optionals is something I really want to avoid, it defeats the whole point of making potentially-empty value visible in the API.